scio-bigquery/src/main/scala/com/spotify/scio/bigquery/client/QueryOps.scala

```diff
@@ -0,0 +1,271 @@
+      val fullJobId = BigQueryUtil.generateJobId(client.project)
+      val jobReference = new JobReference().setProjectId(client.project).setJobId(fullJobId)
+      val job = new Job().setConfiguration(jobConfig).setJobReference(jobReference)
+      client.underlying.jobs().insert(client.project, job).execute()
+    }
+
+    if (dryRun) {
+      dryRunCache.getOrElseUpdate((sqlQuery, flattenResults, useLegacySql), run)
+    } else {
+      run
+    }
+  }
+
+  private def isInvalidQuery(e: GoogleJsonResponseException): Boolean =
+    e.getDetails.getErrors.get(0).getReason == "invalidQuery"
+
+  private[scio] def isLegacySql(sqlQuery: String, flattenResults: Boolean): Boolean = {
+    def dryRun(useLegacySql: Boolean): Try[Job] =
+      run(sqlQuery, null, flattenResults, useLegacySql, dryRun = true)
+
+    sqlQuery.trim.split("\n")(0).trim.toLowerCase match {
+      case "#legacysql"   => true
+      case "#standardsql" => false
+      case _              =>
+        // dry run with SQL syntax first
+        dryRun(false) match {
+          case Success(_)                                                   => false
+          case Failure(e: GoogleJsonResponseException) if isInvalidQuery(e) =>
+            // dry run with legacy syntax next
+            dryRun(true) match {
+              case Success(_) =>
+                Logger.warn(
+                  "Legacy syntax is deprecated, use SQL syntax instead. " +
+                    "See https://cloud.google.com/bigquery/docs/reference/standard-sql/")
+                Logger.warn(s"Legacy query: `$sqlQuery`")
+                true
+              case Failure(f) =>
+                Logger.error(
+                  s"Tried both standard and legacy syntax, query `$sqlQuery` failed for both!")
+                Logger.error("Standard syntax failed due to:", e)
+                Logger.error("Legacy syntax failed due to:", f)
+                throw f
+            }
+          case Failure(e) => throw e
+        }
+    }
+
+  }
+
+  /** Extract tables to be accessed by a query. */
+  def extractTables(sqlQuery: String): Set[TableReference] = {
+    val isLegacy = isLegacySql(sqlQuery, flattenResults = false)
+    val tryJob = run(sqlQuery, null, flattenResults = false, isLegacy, dryRun = true)
+    Option(tryJob.get.getStatistics.getQuery.getReferencedTables) match {
+      case Some(l) => l.asScala.toSet
+      case None    => Set.empty
+    }
+  }
+
+  /** Extract locations of tables to be accessed by a query. */
+  def extractLocation(sqlQuery: String): Option[String] = {
+    val locations = extractTables(sqlQuery)
+      .map(t => (t.getProjectId, t.getDatasetId))
+      .map {
+        case (pId, dId) =>
+          val l = client.underlying.datasets().get(pId, dId).execute().getLocation
+          if (l != null) l else BigQueryConfig.location
+      }
+    require(locations.size <= 1, "Tables in the query must be in the same location")
```

Comment:
should `locations` be collected into a Set before this check? otherwise any query that calls more than one table will automatically fail?

Reply:
@ClaireMcGinty it's a `Set` already 😄